File: src_c/unify_definition.c
#include <math.h>
#include "unify_definition.h"

float getDist(const float a[DIMENTION], const float b[DIMENTION]) {
    float delta = 0;
    float d;
    int i;

    for (i = 0; i < DIMENTION; i++) {
        d = a[i] - b[i];
        delta += d * d;
    }
    return sqrt(delta);
}

void copyPoint(float source[DIMENTION], float dest[DIMENTION]) {
    int i;
    for (i = 0; i < DIMENTION; i++) {
        dest[i] = source[i];
    }    
}

void unifyStroke(float stroke[STROKE_MAX_LENGTH][DIMENTION], float newStroke[SEGMENTATION][DIMENTION], int length) {
    float strokeLengths[STROKE_MAX_LENGTH];
    float step;
    int i, j;
    int newStrokeId = 1;
    float nextLength;
    float * p1;
    float * p2;
    float delta;
    float coeff;

    strokeLengths[0] = 0;

    for (i = 1; i < length; i++) {
        strokeLengths[i] = strokeLengths[i - 1] + getDist(stroke[i - 1], stroke[i]);        
    }

    step = strokeLengths[length - 1] / (SEGMENTATION - 1);

    copyPoint(stroke[0], newStroke[0]);
    copyPoint(stroke[length - 1], newStroke[SEGMENTATION - 1]);

    nextLength = step;

    for (i = 1; i < length; i++) {
        if (strokeLengths[i] > nextLength) {
            p1 = stroke[i - 1];
            p2 = stroke[i];
            delta = strokeLengths[i] - strokeLengths[i - 1];
            coeff = delta == 0 ? 0 : (1 - ((strokeLengths[i] - nextLength) / delta));
            for (j = 0; j < DIMENTION; j++) {
                newStroke[newStrokeId][j] = p1[j] + (p2[j] - p1[j]) * coeff;
            }
            newStrokeId += 1;
            nextLength += step;
            if (newStrokeId == SEGMENTATION - 1) {
                break;                
            }
        }
    }
}

float checkStroke(float stroke[SEGMENTATION][DIMENTION], const float description[SEGMENTATION][DIMENTION + 1]) {
    float errors[SEGMENTATION];
    float mean = 0;
    float result = 0;
    float radius;
    float d;
    int i;

    for (i = 0; i < SEGMENTATION; i++) {
        radius = getDist(stroke[i], description[i]) - description[i][3];
        radius = radius < 0 ? 0 : radius;
        errors[i] = radius;
        mean += radius;
    }

    mean /= SEGMENTATION;

    for (i = 0; i < SEGMENTATION; i++) {
        d = errors[i] - mean;
        result += d * d;
    }

    return sqrt(result / SEGMENTATION);
}

void getLetter(float stroke[STROKE_MAX_LENGTH][DIMENTION], int length, float errors[STROKES_COUNT]) {
    float unifiedStroke[SEGMENTATION][DIMENTION];
    unifyStroke(stroke, unifiedStroke, length);
    int i;
    for (i = 0; i < STROKES_COUNT; i++) {
        errors[i] = checkStroke(unifiedStroke, STROKES[i]);
    }
}

File: src_py/unify_definition.py
#! /usr/bin/env python2.7

import numpy as np


def unify_stroke(stroke, points_count):
    s_delta = np.diff(stroke, axis=0)
    s_length = np.linalg.norm(s_delta, axis=1)
    stroke_lengths = np.cumsum(s_length)

    stroke_lengths = np.hstack((np.array(0), stroke_lengths))

    new_stroke = np.zeros((points_count, stroke.shape[1]))

    new_stroke[0] = stroke[0]
    new_stroke[-1] = stroke[-1]

    step = stroke_lengths[-1] / (points_count - 1)

    next_length = step

    new_stroke_id = 1

    for i in xrange(len(stroke)):
        while (stroke_lengths[i] > next_length):
            p1 = stroke[i - 1]
            p2 = stroke[i]
            delta = stroke_lengths[i] - stroke_lengths[i - 1]
            coeff = (0 if delta == 0
                     else (1 - ((stroke_lengths[i] - next_length) / delta)))

            new_stroke[new_stroke_id] = p1 + (p2 - p1) * coeff
            new_stroke_id += 1
            next_length += step

    return new_stroke


def check_stroke(stroke, description_):
    if not isinstance(description_, np.ndarray):
        description = np.array(description_)
    else:
        description = np.copy(description_)

    delta = stroke - description
    error = np.linalg.norm(delta, axis=1)

    return np.std(error)


def get_letter(stroke, segmentation, letters):
    u_st = unify_stroke(stroke, segmentation)

    result = [(key, check_stroke(u_st, letter))
              for key, letter in letters.items()]

    result.sort(key=lambda x: x[1])

    return result
